### PR TITLE
Syngen vbr

### PR DESCRIPTION
**Pull request by dinkelbachjan**
- The synchronous generator VBR model has been revised and is validated now in the CI.
- An ideal current source model in emt3ph leverages easier validation of synchronous generator models implemented as current source.
- The dimensioning of snubber components for MNA simulation of the transformer models has been revised. It is now based on globally defined constants that specify the snubber power level relative to the component's rated power. This principle has been implemented only for the transformers so far, but could be adopted for other components (e.g. lines).
- An example of the WSCC9 bus system runs now with full order synchronous generator models (DCIM and VBR model).
- The class MNASolverSysRecomp has been removed. Instead, the system recomputation mechanisms have now been integrated into the MNASolver class. Like this, all solver classes inheriting from MNASolver can realize system recomputations by reusing basic mechanisms when implementing methods specific to their data structure or numerical solution approach. This is so far fully implemented for MNASolverEigenSparse.

### CODE REVIEW

File: Include/dpsim/MNASolver.h [hunk 1]
```diff
@@ -93,6 +91,15 @@ namespace DPsim {
 		///
 		std::vector< CPS::Attribute<Matrix>::Ptr > mLeftVectorHarmAttributes;
 
+		// #### MNA specific attributes related to system recomputation
+		/// Number of system matrix recomputations
+		Int mNumRecomputations;
```
Comment (m-mirz) [inline Include/dpsim/MNASolver.h:96]: @dinkelbachjan Is this aggregating the number of recomputations of all time steps?

Comment (dinkelbachjan) [reply Include/dpsim/MNASolver.h:96]: Yes indeed. It was introduced to obtain the overall number of recomputations (refactorisations) performed for a single simulation.

File: Source/MNASolver.cpp [hunk 2]
```diff
@@ -80,7 +80,7 @@ void MnaSolver<VarType>::initialize() {
 	// Some components feature a different behaviour for simulation and initialization
 	for (auto comp : mSystem.mComponents) {
 		auto powerComp = std::dynamic_pointer_cast<CPS::TopologicalPowerComp>(comp);
-		if (powerComp) powerComp->setBehaviour(TopologicalPowerComp::Behaviour::Simulation);
+		if (powerComp) powerComp->setBehaviour(TopologicalPowerComp::Behaviour::MNASimulation);
```
Comment (m-mirz) [inline Source/MNASolver.cpp:83]: @dinkelbachjan Why do we need to differentiate between Simulation and MNASimulation?

File: cmake/GetVillasDPsim.cmake [hunk 3]
```diff
@@ -2,7 +2,7 @@ include(FetchContent)
 FetchContent_Declare(
   dpsim-villas
   GIT_REPOSITORY https://github.com/sogno-platform/dpsim-villas.git
-  GIT_TAG        main
+  GIT_TAG        c14a887531bae6e5067bff39991b78e479003856
```
Comment (m-mirz) [inline cmake/GetVillasDPsim.cmake:5]: @dinkelbachjan this should be set to main again.

File: Source/MNASolverEigenDense.cpp [hunk 4]
```diff
@@ -72,7 +72,6 @@ void MnaSolverEigenDense<Real>::createEmptySystemMatrix() {
 		mSwitchedMatrices[bit].push_back(Matrix::Zero(mNumMatrixNodeIndices, mNumMatrixNodeIndices));
 		mLuFactorizations[bit].push_back(Eigen::PartialPivLU<Matrix>());
 	}
-	mBaseSystemMatrix = Matrix::Zero(mNumMatrixNodeIndices, mNumMatrixNodeIndices);
```
[the remark this answers is not in the document]
Comment (dinkelbachjan) [reply Source/MNASolverEigenDense.cpp:75]: The mechanism remained unchanged. That is, the mechanism on storing the static part for reusing it to build the varying system matrix for faster system recomputation is implemented by `EigenSparse` based on using `mBaseSystemMatrix`

https://github.com/sogno-platform/dpsim/blob/5e36b121b1906f432e782eec807aefbc8e5fee5b/Source/MNASolverEigenSparse.cpp#L113

For `EigenDense`, the system recomputation solver is not implemented yet, as visible for example here
https://github.com/sogno-platform/dpsim/blob/5e36b121b1906f432e782eec807aefbc8e5fee5b/Include/dpsim/MNASolverEigenDense.h#L75

That is why, recomputation specific parts including `mBaseSystemMatrix` I removed so far from `EigenDense`.

File: Source/MNASolver.cpp (hunk 2 above)
Comment (dinkelbachjan) [reply Source/MNASolver.cpp:83]: The `Simulation` behaviour has been removed, while `MNASimulation` and `PFSimulation` have been introduced instead, see

https://github.com/sogno-platform/dpsim/blob/5e36b121b1906f432e782eec807aefbc8e5fee5b/models/Include/cps/TopologicalPowerComp.h#L26

This has been done to be able to be more specific on the type of simulation. In particular, component models from the SP domain are used at the same time for MNA and *standalone* PF simulations. The component model can look different, e.g. snubber components should only be added for MNA simulations, but not for *standalone* PF simulations. An example for that you can find here 

https://github.com/sogno-platform/dpsim/blob/5e36b121b1906f432e782eec807aefbc8e5fee5b/models/Source/SP/SP_Ph1_Transformer.cpp#L129

File: cmake/GetVillasDPsim.cmake (hunk 3 above)
Comment (m-mirz) [reply cmake/GetVillasDPsim.cmake]: I fixed it.

File: Examples/Cxx/Circuits/DP_SynGenTrStab_3Bus_Fault.cpp [hunk 5]
```diff
@@ -205,6 +206,7 @@ void DP_SynGenTrStab_3Bus_Fault(String simName, Real timeStep, Real finalTime, b
 	simDP.setDomain(Domain::DP);
 	simDP.addLogger(loggerDP);
 	simDP.doSystemMatrixRecomputation(true);
+	simDP.setMnaSolverImplementation(MnaSolverFactory::MnaSolverImpl::EigenSparse);
```
Comment (m-mirz) [inline Examples/Cxx/Circuits/DP_SynGenTrStab_3Bus_Fault.cpp:209]: We need to expose this in the pybind interface.

File: Examples/Cxx/Circuits/DP_Slack_PiLine_VSI_Ramp_with_PF_Init.cpp [hunk 6]
```diff
@@ -106,7 +107,7 @@ int main(int argc, char* argv[]) {
 	pv->setParameters(scenario.systemOmega, scenario.pvNominalVoltage, scenario.pvNominalActivePower, scenario.pvNominalReactivePower);
 	pv->setControllerParameters(cmdScaleP*scenario.KpPLL, cmdScaleI*scenario.KiPLL, cmdScaleP*scenario.KpPowerCtrl, cmdScaleI*scenario.KiPowerCtrl, cmdScaleP*scenario.KpCurrCtrl, cmdScaleI*scenario.KiCurrCtrl, scenario.OmegaCutoff);
 	pv->setFilterParameters(scenario.Lf, scenario.Cf, scenario.Rf, scenario.Rc);
-	pv->setTransformerParameters(scenario.systemNominalVoltage, scenario.pvNominalVoltage, scenario.systemNominalVoltage/scenario.pvNominalVoltage, 0, 0, scenario.transformerInductance);
+	pv->setTransformerParameters(scenario.systemNominalVoltage, scenario.pvNominalVoltage, scenario.transformerNominalPower, scenario.systemNominalVoltage/scenario.pvNominalVoltage, 0, 0, scenario.transformerInductance);
```
Comment (m-mirz) [inline Examples/Cxx/Circuits/DP_Slack_PiLine_VSI_Ramp_with_PF_Init.cpp:110]: We could store the scenario parameters in a json file.

**Comment by JTS22** [comment]
On the current commit (aa38963bb183c108bdb7940e536ba5de6b1af60c), the C++ examples DP_EMT_SynGenDq7odODE_ThreePhFault, DP_EMT_SynGenDq7odODE_SteadyState and DP_EMT_SynGenDq7odODE_LoadStep fail to run. When started in a shell, they produce the following output (using DP_EMT_SynGenDq7odODE_SteadyState):

```
[14:42:53.720421 DP_SynGenDq7odODE_SteadyState info] Initialize simulation: DP_SynGenDq7odODE_SteadyState
[14:42:53.720711 MnaSolverFactory info] creating EigenDense solver implementation
[14:42:53.722583 DP_SynGenDq7odODE_SteadyState info] Scheduling tasks.
[14:42:53.723137 DP_SynGenDq7odODE_SteadyState info] Scheduling done.
[14:42:53.723160 DP_SynGenDq7odODE_SteadyState info] Opening interfaces.
[14:42:53.723194 DP_SynGenDq7odODE_SteadyState info] Start synchronization with remotes on 0 interfaces
[14:42:53.723222 DP_SynGenDq7odODE_SteadyState info] Synchronized simulation start with remotes
[14:42:53.723247 DP_SynGenDq7odODE_SteadyState info] Start simulation: DP_SynGenDq7odODE_SteadyState
[14:42:53.723273 DP_SynGenDq7odODE_SteadyState info] Time step: 5.000000e-05
[14:42:53.723293 DP_SynGenDq7odODE_SteadyState info] Final time: 3.000000e-01
[14:42:55.159086 DP_SynGenDq7odODE_SteadyState info] Simulation calculation time: 1.435765
[14:42:55.159407 DP_SynGenDq7odODE_SteadyState info] Simulation finished.
[14:42:55.160397 EMT_SynGenDq7odODE_SteadyState info] Initialize simulation: EMT_SynGenDq7odODE_SteadyState
[14:42:55.160497 MnaSolverFactory info] creating EigenDense solver implementation
[14:42:55.161067 EMT_SynGenDq7odODE_SteadyState info] Scheduling tasks.
[14:42:55.161297 EMT_SynGenDq7odODE_SteadyState info] Scheduling done.
[14:42:55.161309 EMT_SynGenDq7odODE_SteadyState info] Opening interfaces.
[14:42:55.161318 EMT_SynGenDq7odODE_SteadyState info] Start synchronization with remotes on 0 interfaces
[14:42:55.161324 EMT_SynGenDq7odODE_SteadyState info] Synchronized simulation start with remotes
[14:42:55.161332 EMT_SynGenDq7odODE_SteadyState info] Start simulation: EMT_SynGenDq7odODE_SteadyState
[14:42:55.161342 EMT_SynGenDq7odODE_SteadyState info] Time step: 5.000000e-05
[14:42:55.161354 EMT_SynGenDq7odODE_SteadyState info] Final time: 3.000000e-01
DP_EMT_SynGenDq7odODE_SteadyState: /usr/include/eigen3/Eigen/src/Core/DenseCoeffsBase.h:364: Eigen::DenseCoeffsBase<Derived, 1>::Scalar& Eigen::DenseCoeffsBase<Derived, 1>::operator()(Eigen::Index, Eigen::Index) [with Derived = Eigen::Matrix<double, -1, -1>; Eigen::DenseCoeffsBase<Derived, 1>::Scalar = double; Eigen::Index = long int]: Assertion `row >= 0 && row < rows() && col >= 0 && col < cols()' failed.
Aborted (core dumped)
```

This behavior is mirrored in the corresponding Python notebooks SynGenDq7od_SteadyState_DP_EMT and SynGenDq7od_ThreePhFault_DP_EMT where the Jupyter Kernel crashes upon running the EMT simulation for the ODE model.

**Comment by JTS22** [comment]
https://github.com/sogno-platform/dpsim/pull/48#issuecomment-971656944
Similar behavior can be observed for the SP_SynGenTrStab_3Bus_Fault Cpp example and corrsponding DP_SP_SynGenTrStab_3Bus_Fault notebook:
```
[root@1b13b00c56aa Cxx]# ./SP_SynGenTrStab_3Bus_Fault 
[10:48:06.748377 PiLine12 warning] Zero value for Conductance, setting default value of G=1e-06 [S]
[10:48:06.748772 PiLine13 warning] Zero value for Conductance, setting default value of G=1e-06 [S]
[10:48:06.749126 PiLine23 warning] Zero value for Conductance, setting default value of G=1e-06 [S]
[10:48:06.750046 SP_SynGenTrStab_3Bus_Fault_PF info] Initialize simulation: SP_SynGenTrStab_3Bus_Fault_PF
[10:48:06.750667 SP_SynGenTrStab_3Bus_Fault_PF_PF_Solver warning] Unable to get base voltage at n3
[10:48:06.751197 SP_SynGenTrStab_3Bus_Fault_PF info] Scheduling tasks.
[10:48:06.751516 SP_SynGenTrStab_3Bus_Fault_PF info] Scheduling done.
[10:48:06.751534 SP_SynGenTrStab_3Bus_Fault_PF info] Opening interfaces.
[10:48:06.751548 SP_SynGenTrStab_3Bus_Fault_PF info] Start synchronization with remotes on 0 interfaces
[10:48:06.751564 SP_SynGenTrStab_3Bus_Fault_PF info] Synchronized simulation start with remotes
[10:48:06.751577 SP_SynGenTrStab_3Bus_Fault_PF info] Start simulation: SP_SynGenTrStab_3Bus_Fault_PF
[10:48:06.751607 SP_SynGenTrStab_3Bus_Fault_PF info] Time step: 3.000000e+01
[10:48:06.751645 SP_SynGenTrStab_3Bus_Fault_PF info] Final time: 6.000000e+01
[10:48:06.754569 SP_SynGenTrStab_3Bus_Fault_PF info] Simulation calculation time: 0.002899
[10:48:06.754801 SP_SynGenTrStab_3Bus_Fault_PF info] Simulation finished.
[10:48:06.755959 PiLine12 warning] Zero value for Conductance, setting default value of G=1e-06 [S]
[10:48:06.756276 PiLine13 warning] Zero value for Conductance, setting default value of G=1e-06 [S]
[10:48:06.756506 PiLine23 warning] Zero value for Conductance, setting default value of G=1e-06 [S]
[10:48:06.757521 SP_SynGenTrStab_3Bus_Fault_SP info] Initialize simulation: SP_SynGenTrStab_3Bus_Fault_SP
[10:48:06.757722 MnaSolverFactory info] creating EigenDense solver implementation
terminate called after throwing an instance of 'CPS::SystemError'
Aborted (core dumped)
```

**Pull request merged**